testing/setup_test.go

```diff
@@ -0,0 +1,186 @@
+package testing
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"github.com/testcontainers/testcontainers-go"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	chainid = "horcux"
+	netid   = "horcrux"
+)
+
+// disable logging from testcontainers
+func init() {
+	log.Default().SetOutput(ioutil.Discard)
+}
+
+func TestUpgradeValidatorToHorcrux(t *testing.T) {
+	testsDone := make(chan struct{})
+	contDone := make(chan struct{})
+	home, err := ioutil.TempDir("", "")
+	require.NoError(t, err)
+
+	ctx := context.Background()
+	provider, err := testcontainers.NewDockerProvider()
+	require.NoError(t, err)
+
+	net, err := provider.CreateNetwork(ctx, testcontainers.NetworkRequest{Name: netid, Internal: false})
+	require.NoError(t, err)
+
+	nodes := MakeTestNodes(4, home, chainid, simdChain, provider)
+
+	cont := startValidatorContainers(t, provider, net, nodes)
+	require.NoError(t, err)
+	nodes.LogGenesisHashes(t)
+
+	// set the test cleanup function
+	go cleanUpTest(t, testsDone, contDone, cont, net, home)
+	t.Cleanup(func() {
+		testsDone <- struct{}{}
+		<-contDone
+	})
+
+	for i, n := range nodes {
+		str, err := cont[i].PortEndpoint(ctx, "26657", "http")
+		require.NoError(t, err)
+		fmt.Printf("%s available at %s\n", n.Name(), str)
+		require.NoError(t, n.NewClient(str))
+	}
+
+	time.Sleep(10 * time.Second)
```

Comment:
Why sleep here? Are you waiting for containers to start? I'm still learning the testcontainers package (which looks amazing) and think that their wait package might be helpful: github.com/testcontainers/testcontainers-go/wait